cli/tools/upgrade.rs

```diff
@@ -11,16 +13,90 @@ use deno_runtime::deno_fetch::reqwest::Client;
 use once_cell::sync::Lazy;
 use std::env;
 use std::fs;
+use std::io::Read;
+use std::io::Seek;
 use std::io::Write;
 use std::path::Path;
 use std::path::PathBuf;
 use std::process::Command;
+use std::time::Duration;
+use std::time::SystemTime;
 
 static ARCHIVE_NAME: Lazy<String> =
   Lazy::new(|| format!("deno-{}.zip", env!("TARGET")));
 
 const RELEASE_URL: &str = "https://github.com/denoland/deno/releases";
 
+const UPGRADE_CHECK_INTERVAL: Duration = Duration::from_secs(12 * 60 * 60);
+
+const UPGRADE_CHECK_FETCH_DELAY: Duration = Duration::from_millis(500);
+
+pub fn check_for_upgrades() -> Option<String> {
+  let current_exe_mtime =
+    env::current_exe().ok()?.metadata().ok()?.modified().ok()?;
+
+  // Open `$DENO_DIR/latest.txt` for reading and writing. If this fails somehow
+  // we bail out early and return `None`.
+  let latest_txt_path = DenoDir::new(None).ok()?.root.join("latest.txt");
+  let mut latest_txt_file = fs::OpenOptions::new()
+    .create(true)
+    .read(true)
+    .write(true)
+    .open(&latest_txt_path)
+    .ok()?;
+  let latest_txt_mtime = latest_txt_file.metadata().ok()?.modified().ok()?;
+  let latest_txt_age = SystemTime::now()
+    .duration_since(latest_txt_mtime)
+    .unwrap_or_default();
```

Comment:
There are around 6 syscalls _just_ here. `readlink`, `stat`, `open`, `fstat`, `clock_gettime` & a `read` below.

And then atleast 2 syscalls in `DenoDir::new(None)`.
https://github.com/denoland/deno/blob/7a1a082876298a4c9e37237074ea62942180d083/cli/deno_dir.rs#L32-L44